src/Adapter/MSTest.TestAdapter/Helpers/MethodRunner.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Globalization;
+using System.Reflection;
+
+using Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.ObjectModel;
+
+using UnitTestOutcome = Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.ObjectModel.UnitTestOutcome;
+
+namespace Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Helpers;
+
+internal static class MethodRunner
+{
+    internal static TestFailedException? RunWithTimeoutAndCancellation(
+        Action action, CancellationTokenSource cancellationTokenSource, int? timeout, MethodInfo methodInfo,
+        string methodCancelledMessageFormat, string methodTimedOutMessageFormat)
+    {
+        try
+        {
+            var executionTask = Task.Run(action, cancellationTokenSource.Token);
+
+            if (!timeout.HasValue)
+            {
+                executionTask.Wait(cancellationTokenSource.Token);
+                return cancellationTokenSource.IsCancellationRequested
```

Comment:
Wonder if this make sense...if we reach this point means that anyway the test ran...and we have a race with a cancellation but "we have the result" actually, so we could return null, something else in-progress eventually fire some OCE

Reply:
Yep looks like it always throws so I should return only `null`.